### DISCUSSION TITLE
v3.1.2

### DISCUSSION THREAD

**Discussion by danielroe** [nuxt/nuxt#18741] [Announcements]
> **3.1.2** is a patch release with bug fixes (particularly focusing on performance and DX).

## Changelog

[compare changes](https://github.com/nuxt/nuxt/compare/v3.1.1...v3.1.2)

### 🔥 Performance

  - **nuxt:** Simplify generated variable names ([#18629](https://github.com/nuxt/nuxt/pull/18629))
  - **vite:** Use compiled regexp for test ([#18646](https://github.com/nuxt/nuxt/pull/18646))
  - **nuxt:** Cache result of importing styles module ([#18734](https://github.com/nuxt/nuxt/pull/18734))

### 🩹 Fixes

  - **nuxt:** Provide fallback values for undefined runtime config ([#18586](https://github.com/nuxt/nuxt/pull/18586))
  - **kit:** Don't use default export of defu ([#18589](https://github.com/nuxt/nuxt/pull/18589))
  - Use named export from `defu` in all places ([#18624](https://github.com/nuxt/nuxt/pull/18624))
  - **nuxt:** Dedupe vue-router ([#18626](https://github.com/nuxt/nuxt/pull/18626))
  - **test-utils:** Prevent orphaned processes and use baseURL when loading ([#18623](https://github.com/nuxt/nuxt/pull/18623))
  - **vite:** Ensure `__publicAssetsURL` set before loading assets ([#18642](https://github.com/nuxt/nuxt/pull/18642))
  - **kit:** Avoid adding already installed modules to internal `_installedModules` ([#18647](https://github.com/nuxt/nuxt/pull/18647))
  - **nuxt:** Make `onNuxtReady` safe to run on server-side ([#18706](https://github.com/nuxt/nuxt/pull/18706))
  - **vite, webpack:** Omit magic keys when import of same name is detected ([#18733](https://github.com/nuxt/nuxt/pull/18733))

### 📖 Documentation

  - Update `vue-gtag` plugin example ([#18528](https://github.com/nuxt/nuxt/pull/18528))
  - Mention virtual file system ([#18546](https://github.com/nuxt/nuxt/pull/18546))
  - Mention head composable as alternative to `useHead` ([#18552](https://github.com/nuxt/nuxt/pull/18552))
  - Use `defineEventHandler()` to avoid warnings ([#18557](https://github.com/nuxt/nuxt/pull/18557))
  - Remove unnecessary `JSON.stringify()` ([#18590](https://github.com/nuxt/nuxt/pull/18590))
  - Updated unreachable docus url in documentation ([#18618](https://github.com/nuxt/nuxt/pull/18618))
  - Add link to pages documentation to routing ([#18602](https://github.com/nuxt/nuxt/pull/18602))
  - Add comment about needing to install `@types/node` manually ([6b2bc680b](https://github.com/nuxt/nuxt/commit/6b2bc680b))
  - Add note about directory import ([d2c00dc46](https://github.com/nuxt/nuxt/commit/d2c00dc46))
  - Fix array syntax ([f14f3815f](https://github.com/nuxt/nuxt/commit/f14f3815f))
  - Fix buttons width on getting started ([#18643](https://github.com/nuxt/nuxt/pull/18643))
  - Add `.env` to directory structure and improve config docs ([#18594](https://github.com/nuxt/nuxt/pull/18594))
  - Mention options api equivalent for `head()` ([#18650](https://github.com/nuxt/nuxt/pull/18650))
  - Fix broken link to esm section ([#18716](https://github.com/nuxt/nuxt/pull/18716))
  - Improve routing `validate` example ([#18728](https://github.com/nuxt/nuxt/pull/18728))

### 🏡 Chore

  - Update `2.x` branch name ([727cf7958](https://github.com/nuxt/nuxt/commit/727cf7958))

### ✅ Tests

  - **nuxt:** Exclude new internal vue `assertNumber` helper ([aa646f065](https://github.com/nuxt/nuxt/commit/aa646f065))

### 🤖 CI

  - Publish nightly `nuxt-edge` for nuxt v2 ([dd0e2643c](https://github.com/nuxt/nuxt/commit/dd0e2643c))
  - Fix ref for 2.x branch ([21badd9f2](https://github.com/nuxt/nuxt/commit/21badd9f2))
  - Allow manually dispatching nightly nuxt2 build ([09114125c](https://github.com/nuxt/nuxt/commit/09114125c))

### ❤️  Contributors

- Daniel Roe <daniel@roe.dev>
- Aimer-fan 
- Rémi Robichet <remi.robichet@gmail.com>
- Pooya Parsa <pyapar@gmail.com>
- Florent Delerue <florentdelerue@hotmail.com>
- Tobias Diez <code@tobiasdiez.com>
- Jenni <jenni@ohsomodern.org>
- Arun Govil <a.arungovil@gmail.com>
- 华晨 <hclj37@gmail.com>
- Inesh Bose 
- Theo Ephraim <theozero@gmail.com>

<hr /><em>This discussion was created from the release <a href='https://github.com/nuxt/nuxt/releases/tag/v3.1.2'>v3.1.2</a>.</em>

**Comment by falsisdev** [comment]
Great! I hope the [`loading`](https://nuxt.com/docs/migration/component-options#loading) component will be added to version `3.x` soon!

**Reply by danielroe** [reply]
Oh! It is, actually supported already: https://nuxt.com/docs/api/components/nuxt-loading-indicator#nuxtloadingindicator. But we need to update the migration guide.

**Comment by dosstx** [comment]
Ran `npx nuxi upgrade` and get this error:

```
npm WARN EBADENGINE Unsupported engine {
npm WARN EBADENGINE   package: 'rollup@3.13.0',
npm WARN EBADENGINE   required: { node: '>=14.18.0', npm: '>=8.0.0' },
npm WARN EBADENGINE   current: { node: 'v18.14.0', npm: '7.18.1' }
```

**Reply by DrJume** [reply]
Try to upgrade npm, as stated, to version 8. You can do this with `npm -g install npm`

**Reply by dosstx** [reply]
Thanks! Works!

**Comment by lehni** [comment]
What a great release! `v3.1.2` just halved the startup time required for `nuxi dev` for us. `nuxi build` has also become much faster. Thanks for the great work!

**Comment by mcrapts** [comment]
Good to see the pace going! I'm now facing a bug with Nuxt Content though. I filed an [issue](https://github.com/nuxt/content/issues/1884) in the Nuxt Content repository, but perhaps it's related to Nuxt itself?

**Comment by ChuanfengLai** [comment]
After `npx nuxt upgrade --force`，run `npm start`，ERROR 
```
[nuxt] [request error] [unhandled] [500] require is not defined in ES module scope, you can use import instead                                                                                                                                                                                                    11:11:19
  at $id_3vCgCbt7m0 (/D:/project/xlco_ssr_content/.nuxt/dist/server/server.mjs:552:3)
  at __instantiateModule__ (/D:/project/xlco_ssr_content/.nuxt/dist/server/server.mjs:165645:9)
  at __ssrLoadModule__ (/D:/project/xlco_ssr_content/.nuxt/dist/server/server.mjs:165583:25)
  at ssrImport (/D:/project/xlco_ssr_content/.nuxt/dist/server/server.mjs:165608:13)
  at $id_w4M3WWtjnW (/D:/project/xlco_ssr_content/.nuxt/dist/server/server.mjs:534:37)
  at __instantiateModule__ (/D:/project/xlco_ssr_content/.nuxt/dist/server/server.mjs:165645:9)
  at __ssrLoadModule__ (/D:/project/xlco_ssr_content/.nuxt/dist/server/server.mjs:165583:25)
  at ssrImport (/D:/project/xlco_ssr_content/.nuxt/dist/server/server.mjs:165608:13)
  at $id_UrXgbFYuQs (/D:/project/xlco_ssr_content/.nuxt/dist/server/server.mjs:17:37)
  at __instantiateModule__ (/D:/project/xlco_ssr_content/.nuxt/dist/server/server.mjs:165645:9)
```

**Reply by danielroe** [reply]
Please feel free to open an issue with a [reproduction](https://nuxt.com/docs/community/reporting-bugs/#create-a-minimal-reproduction).

**Comment by rahulkumarsingh73690** [comment]
ERROR  Cannot start nuxt:  Cannot find module 'D:\Programming\node_modules\unstorage\dist\drivers\utils\node-fs' imported from D:\Programming\node_modules\unstorage\dist\drivers\fs.mjs

**Reply by pi0** [reply]
temporary issue fixed in latest unstorage. please try refrashing lockfile using `npx nuxi upgrade --force`.
